Add CDI devices to device plugin API

elezar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

This PR adds support for specifying CDI device IDs in the Device Plugin API. This allows plugin authors to uniquely specify CDI devices which are passed to the CRI runtimes in a dedicated field instead of relying on annotations or enviroment variables.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
See enhancement issue https://github.com/kubernetes/enhancements/issues/4009

#### Special notes for your reviewer:

This is related to the changes implemented in https://github.com/kubernetes/kubernetes/pull/115891, and uses the same field for standard Device Plugins.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
A CDIDevice field is includes in the Device Plugin's `ContainerAllocateResponse`. This field maps to the CDIDevice field in the CRI protocol.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [Usage]: <link>
- [Other doc]: <link>
-->
```doc
- [KEP]: https://github.com/kubernetes/enhancements/issues/4009
```

elezar committed:
Add CDI devices to device plugin API

This change adds CDI device IDs to the ContainerAllocateResponse in the
device plugin API. This allows a device plugin to specify CDI devices
by their unique fully-qualified CDI device names using the related field
in the CRI specification.

Signed-off-by: Evan Lezar <elezar@nvidia.com>

ffromani commented:
/ok-to-test

ffromani commented:
/triage accepted

per linked KEP

bart0sh commented:
/priority important-longterm
/cc

bart0sh commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -321,6 +323,20 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
+			if _, ok := cdiDevicesMap[cdiDevice.Name]; ok {
+				klog.V(4).InfoS("Skip existing CDI Device", "name", cdiDevice.Name)
+				continue
+			}
```
Is cdiDeviceMap update anywhere? If not, how can it contain existing CDI devices?

elezar replied:
It's updated on line 334 of this change set. It was missing off the first revision though.

bart0sh replied:
it makes sense to me now, thanks.

bart0sh commented:
/cc @swatisehgal @ffromani

bart0sh commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -244,6 +244,7 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
 	mountsMap := make(map[string]string)
 	envsMap := make(map[string]string)
 	annotationsMap := make(map[string]string)
+	cdiDevicesMap := make(map[string]string)
```
```suggestion
	cdiDevicesMap := make(map[string]struct{})
```

bart0sh commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -321,6 +323,20 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
+				continue
+			}
+			klog.V(4).InfoS("Add CDI device", "name", cdiDevice.Name)
+			cdiDevicesMap[cdiDevice.Name] = cdiDevice.Name
```
```suggestion
			cdiDevicesMap[cdiDevice.Name] = struct{}
```

bart0sh commented:
/retest

bart0sh commented on staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/api.proto:
```diff
@@ -185,6 +194,8 @@ message ContainerAllocateResponse {
 	map<string, string> annotations = 4;
+	// CDI devices for the container.
+	repeated CDIDevice cdi_devices = 5;
 }
```
This is an API change as far as I understand. Did you test if current/old plugins still work with it?

elezar replied:
I need to still dig through the tests for the device plugin. I will update the PR with details once I have better insight into this.

bart0sh replied:
It would be good to cover this functionality with unit and e2e tests.

elezar replied:
Yes, I had a look at the existing unit tests today and they would require some work before adding new ones would be simple. I am also looking into the e2e tests and will definitely add them there.

ffromani replied:
I also agree (but it seems there's general consensus already) adding unit tests would be nice. Probably the best way bar only way here is extending `sample-device-plugin`. We need to document we need to keep variants with/without `CDIDevices` reporting until the feature is GA. Will require some coordination but is doable.
I think we should start the e2e work and call it a prerequisite for the beta graduation but I don't think it should block alpha. I recon setting up the bits and pieces to enable e2e testing will take some time.

klueska replied:
It is not an API change in the sense that we need a version bump. Additional fields added to a protobuf don't break the API for components that are built against an earlier revision of the proto spec without the additional field.

elezar committed:
Add a builder for ContainerAllocateResponse objects

This chagne introduces a helper to construct ContainerAllocateResponse instances.
Test cases are updated to use a new constructor accepting functional options
allowing the response contents to be set based on the test requirements.

This can then be extended to also test additional fields in the device plugin API
such as annotations which are not currently covered or new fields.

Signed-off-by: Evan Lezar <elezar@nvidia.com>

haircommander commented:
/hold

on consensus in https://github.com/kubernetes/enhancements/pull/4011

elezar committed:
Move CDI annotation code to utils package

Signed-off-by: Evan Lezar <elezar@nvidia.com>

elezar committed:
Generate empty cdi annotations

Signed-off-by: Evan Lezar <elezar@nvidia.com>

ffromani commented on pkg/kubelet/cm/util/cdi/cdi.go:
```diff
@@ -39,8 +39,8 @@ const (
-// generate container annotations using CDI UpdateAnnotations API.
-func generateCDIAnnotations(
+// GenerateCDIAnnotations generate container annotations using CDI UpdateAnnotations API.
+func GenerateCDIAnnotations(
```
minor/nonblocking: if we are creating a package called `cdi`, we can rename `GenerateAnnotations` so it reads
`cdi.GenerateAnnotations` which is nicer.

klueska replied:
+1

elezar replied:
I have updated this.

ffromani commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -244,14 +249,16 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
 	mountsMap := make(map[string]string)
 	envsMap := make(map[string]string)
 	annotationsMap := make(map[string]string)
+	cdiDevicesMap := make(map[string]struct{})
```
using a `struct{}` as value in a map suggest this is actually a set. Is that the case?

elezar replied:
This could be condidered a set, I suppose. Note that the only intent of this variable is to ensure that we de-duplicate CDI devices. Since these are not key-value pairs, there are no other values that could possibly be overwritten. Let me see if switching to a set cleans things up at all.

ffromani replied:
Yup. My comment is nonblocking for alpha anyway, we still can improve later.

klueska replied:
A string set would probably be nicer here.

elezar replied:
I have used a `Set[string]` here instead. I don't think it cleans things up too much, but checking `.Has` is at least a little easier to reason about than the map semantics.

ffromani commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -321,10 +328,70 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
+
+	var cdiDeviceNames []string
+	for _, cdiDevice := range responseCDIDevices {
+		if _, ok := cdiDevicesMap[cdiDevice.Name]; ok {
+			klog.V(4).InfoS("Skip existing CDI Device", "name", cdiDevice.Name)
+			continue
+		}
+		klog.V(4).InfoS("Add CDI device", "name", cdiDevice.Name)
+		cdiDevicesMap[cdiDevice.Name] = struct{}{}
+
+		cdiDeviceNames = append(cdiDeviceNames, cdiDevice.Name)
```
*IFF* `cdiDevicesMap` is actually behaving like a set, and if we switch to `apimachinery/util/set` I think we will be able to simplify this code. Worth a shot.

elezar replied:
Using a `Set` here would only solve for local (per-resource) uniqueness and not global uniqueness. Maybe that is an indication that using a set is not what we want and we should store the resource name in our `cdiDevicesMap` to provide logs as to which resource is also requesting a specific CDI device.

ffromani commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -321,10 +328,70 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
+		cdiDevices = append(cdiDevices, device)
+	}
+
+	annotationUID := strings.Join([]string{podUID, resourceName}, "-")
```
this works, but a simple string concatenation (`annotationUID := podUID + "-" + resourceName`) seem both simpler and clearer unless we plan to Join more identifiers very very soon

elezar replied:
I have updated this to use standard concatenation -- although I also include the contianer name in the ID now. This may not be required since these annotations are per-container by definition.

ffromani commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -321,10 +328,70 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
+	}
+
+	var annotations []kubecontainer.Annotation
+	for _, a := range cdiAnnotations {
+		if e, ok := annotationsMap[a.Name]; ok {
+			klog.V(4).InfoS("Skip existing annotation", "annotationKey", a.Name, "annotationValue", a.Value)
+			if e != a.Value {
+				klog.ErrorS(nil, "Annotation has conflicting setting", "annotationKey", a.Name, "expected", e, "got", a.Value)
+			}
+			continue
+		}
+		klog.V(4).InfoS("Add annotation", "annotationKey", a.Name, "annotationValue", a.Value)
+		annotationsMap[a.Name] = a.Value
+		annotations = append(annotations, kubecontainer.Annotation{Name: a.Name, Value: a.Value})
+	}
```
I like the shorter possible names, but I reluctantly moved away from single-letter identifiers: `e` and `a` are possibly too short. Can we pick a short-but-a-bit-longer name, also if possible easier to `grep`?

elezar replied:
I have updated this. Now `cdiAnnotations` is the slice that we are constructing and I'm using just `annotations` and `annotations` as the loop variables.

ffromani commented on test/images/sample-device-plugin/sampledeviceplugin.go:
```diff
@@ -66,6 +66,12 @@ func stubAllocFunc(r *pluginapi.AllocateRequest, devs map[string]pluginapi.Devic
 				ContainerPath: fpath,
 				HostPath:      fpath,
 			})
+
+			// add the CDI device ID to the response.
+			cdiDevice := &pluginapi.CDIDevice{
+				Name: fmt.Sprintf("cdi.%s=%s", resourceName, dev.ID),
+			}
+			response.CdiDevices = append(response.CdiDevices, cdiDevice)
```
yes, this is a good start in the right direction. I think we should also:
1. make it possible for the plugin to optionally return empty data (depending on e2e test scenario)
2. keep and use a old version of the plugin without the field at all - this can be done just keeping the old `sample-device-plugin` image in the k8s registry, which should be the case by default

With these two changes in places we should be enabled to test all the major scenarios e2e.

elezar replied:
Yes, I started looking at the e2e testing and noted that quite a bit of work is required to cover the expected use cases for proper coverage. For now, I thought I would add this field to simulate a device plugin that supports this field communicating with a kubelet that doesn't yet support it -- or has the feature turned off.

On the other hand, it may make more sense to do this in a follow-up that specifically targets e2e testing instead of including this here. Do you have thoughts there?

ffromani replied:
I think is better to do everything in a separate (set of) PR(s).
In general I think e2e tests are _desirable_, but not required at this stage. I think the best way forward is focus on good unit test coverage, file an issue or any form of tracking about adding e2e tests for this feature and point this out in the PR. Adding e2e test should be a Beta blocker, and we should keep working on this in this cycle or in the next one, alongside the other activities. 
Let's see what other reviewers think and if this path forward is reasonnable.

elezar replied:
OK. Let me roll-back the e2e testing changes then as part of this PR and then start looking at follow ups. I have added some better unit test coverage with the last iteration.

ffromani reviewed (commented):
looks generally good, mostly minorish questions/suggestions inside

klueska commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -615,25 +675,40 @@ func TestCheckpoint(t *testing.T) {
+	)
 	testManager.podDevices.insert("pod3", "con3", resourceName3,
 		checkpoint.DevicesPerNUMA{nodeWithoutTopology: []string{"dev5"}},
-		constructAllocResp(map[string]string{"/dev/r1dev5": "/dev/r1dev5"},
-			map[string]string{"/home/r1lib1": "/usr/r1lib1"}, map[string]string{}))
+		newContainerAllocateResponse(
+			withCDIDevices("domain2.com/resource3=dev5"),
+		),
+	)
```
Did we lose the original test here?

elezar replied:
I don't think so. The other permutations of data being restored from the checkpoint are tested by the other resources. I can readd them here though. I don't think it was strictly speaking necessary to remove them.

Note that this resource references the same mounts as dev1, and I think this is an oversight from the original tests. @ffromani do you know whether this is the case?

elezar replied:
Actually, I think we may have changed the test here since we're not including a device now. I will revert.

klueska replied:
Did you miss doing this in your latest revision?

klueska commented on staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/api.proto:
```diff
@@ -185,6 +194,8 @@ message ContainerAllocateResponse {
 	// Container annotations to pass to the container runtime
 	map<string, string> annotations = 4;
+	// CDI devices for the container.
+	repeated CDIDevice cdi_devices = 5;
```
Let's define this as follows so that the go code interprets this as CDIDevices, not CdiDevices:
```
cdi_devices = 5 [(gogoproto.customname) = "CDIDevices"];
```

elezar replied:
OK. I can update. For what it's worth, I was just following what was done for the CRI spec, but I prefer the capitilized version and will update.

klueska commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -321,10 +328,70 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
+			cdiDevicesMap,
+		)
+		opts.CDIDevices = append(opts.CDIDevices, cdiDevices...)
+		opts.Annotations = append(opts.Annotations, cdiAnnotations...)
```
Do we need to dedup the annotations here? Is it possible that someone has set the annotations exactly the same as what `getCDIDeviceInfo()` will set. If so, does it matter if they are duplicated in the `opts.Annotations` slice?

**Edit**, I see you do that below with:
```
			klog.V(4).InfoS("Skip existing annotation", "annotationKey", a.Name, "annotationValue", a.Value)
			if e != a.Value {
				klog.ErrorS(nil, "Annotation has conflicting setting", "annotationKey", a.Name, "expected", e, "got", a.Value)
			}
			continue
```

Can we pull the deduplication up to this level rather than burying it inside `getCDIDeviceInfo()` I feel like it would be clearer that way. It would also clean up the api to  `getCDIDeviceInfo()` since you would now drop the following 2:
```
			annotationsMap,
			cdiDevicesMap,
````
(since I'm assuming the only reason you passed an empty cdiDevicesMap in in the first place was to make the API look more symmetrical).

elezar replied:
I think the likelihood of collisions in the annotations is pretty low, but since we can't rule them out entirely, I have left the warnings / errors in.

I don't agree that the `cdiDevicesMap` is there just to make the API symetrical. We also deduplicate the CDI devices requested at this level which requires the device map. Note that it's only empty at the first call after which it will contain all CDI devices for the first ContainerAllocateResponse processed.

klueska replied:
Ah, right -- because its in a loop.

klueska replied:
If deduplication is happening in both cases, then keeping it inside the `getCDIDeviceInfo` call makes sense.

klueska commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -321,10 +328,70 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
 
+// getCDIDeviceInfo returns CDI devices and annotations for a given pod.
+func getCDIDeviceInfo(podUID, resourceName string, responseCDIDevices []*pluginapi.CDIDevice, annotationsMap map[string]string, cdiDevicesMap map[string]struct{}) ([]kubecontainer.CDIDevice, []kubecontainer.Annotation) {
+	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DevicePluginCDIDevices) {
+		return nil, nil
+	}
```
In general, we try and make the calls to the check the feature gate at the highest level possible, rather than buried within function calls that "do nothing" if the feature gate is disabled. In this case, that would mean pulling the feature gate up into pod_device.go and wrapping the following with it:

```
		// Updates for CDI devices. This also updates the annotations.
		cdiDevices, cdiAnnotations := getCDIDeviceInfo(
			podUID, resourceName,
			resp.CdiDevices,
			annotationsMap,
			cdiDevicesMap,
		)
		opts.CDIDevices = append(opts.CDIDevices, cdiDevices...)
```

elezar replied:
OK. I will update.

klueska commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -321,10 +328,70 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
+		return cdiDevices, nil
+	}
+
+	var annotations []kubecontainer.Annotation
+	for _, a := range cdiAnnotations {
+		if e, ok := annotationsMap[a.Name]; ok {
+			klog.V(4).InfoS("Skip existing annotation", "annotationKey", a.Name, "annotationValue", a.Value)
+			if e != a.Value {
+				klog.ErrorS(nil, "Annotation has conflicting setting", "annotationKey", a.Name, "expected", e, "got", a.Value)
+			}
+			continue
+		}
+		klog.V(4).InfoS("Add annotation", "annotationKey", a.Name, "annotationValue", a.Value)
+		annotationsMap[a.Name] = a.Value
+		annotations = append(annotations, kubecontainer.Annotation{Name: a.Name, Value: a.Value})
+	}
```
As mentioned previously, lets pull this out to its own function and called outside of this function.

klueska commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -321,10 +327,69 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
+
+		// Updates for CDI devices. This also updates the annotations.
+		cdiDevices, cdiAnnotations := getCDIDeviceInfo(
+			types.UID(podUID+"-"+contName+"-"+resourceName),
```
Is this just something unique (i.e. it doesn't matter exactly what this is as long as it's unique)?

elezar replied:
Yes, it just needs to be unique per container.

elezar commented:
@klueska @ffromani thanks for the round of reviews. I have updated the implementation following your feedback and would appreciate another pass.

elezar commented:
/test pull-kubernetes-unit

elezar commented:
Then failing tests are in `k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage: cacher`, and I would assume that these are unrelated. Are there known issues with these at the moment?

klueska commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -321,10 +328,69 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
+	// Although the CDI devices are expected to be empty when this feature is disabled, we still
+	// guard this with a feature gate to avoid any potential issues.
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DevicePluginCDIDevices) {
+		cdiAnnotations := getCDIAnnotations(podUID+"-"+contName, sets.List[string](allCDIDevices), annotationsMap)
```
I'm still hung up on this resourceID of `"podUID+"-"+contName`. What is the annotation that will be generated from this? Do we as an end user even care what this, as long as the machinery below knows how to interpret this and inject the proper devices? Can we avoid passing this in and just generate a unique uuid within the body of `getCDIAnnotations`?

elezar replied:
The unit tests show the annotations that are generated:
```
cdi.k8s.io/devicemanager_podUID-contName
```
The container runtimes such as containerd or crio that would interpret these annotations do not parse these fields and only look at the `cdi.k8s.io/` prefix and the values to get the list of devices. The ID is only needed to ensure that there are no collisions at a container level.

If you think a UUID is better, I'm happy to use that instead, but would then ask for guidance on how this can be generated deterministically for testing purposes.

ffromani replied:
For the record I have no strong opinions here and I'll be OK with whatever @klueska (who knows the CDI area in general much better) prefers

klueska commented on pkg/kubelet/cm/devicemanager/pod_devices.go:
```diff
@@ -321,10 +328,69 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
+	// Although the CDI devices are expected to be empty when this feature is disabled, we still
+	// guard this with a feature gate to avoid any potential issues.
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DevicePluginCDIDevices) {
+		cdiAnnotations := getCDIAnnotations(podUID+"-"+contName, sets.List[string](allCDIDevices), annotationsMap)
```
You can call `allDvices.UnsortedList()` instead of casting to a `sets.List[string]`. In fact, maybe its better to do this within the body of the function to keep this callsite more compact.

elezar replied:
I explicitly use `sets.List` to get a sorted list here. One motivation is for odering of the devices in the annotations in the tests.

I can move to a local variable for compactness though.

klueska replied:
Let's leave the resource ID as is for now (I'm still not a huge fan, but it's not worth bikeshedding over) -- just please add a comment as to what its purpose is.

After that and the change to pull sets.List[string] into the function, I think this PR is good to go.

elezar replied:
Updated in https://github.com/kubernetes/kubernetes/commit/1d21aa4cc0f7d94463fe867aadf4608317ffbaf0

ffromani reviewed (commented):
note about API review: https://github.com/kubernetes/kubernetes/pull/118254#issuecomment-1563121596 this is a node specific API and historically the API reviewers tend to defer this special form of API review to SIG-node members, so I think we don't need extra reviews in this area.

ffromani commented:
> Then failing tests are in `k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage: cacher`, and I would assume that these are unrelated. Are there known issues with these at the moment?

It's most likely a flake. 99 out of 100 not blocking this PR. We (as project) should track and fix it, that's it.

ffromani commented:
@elezar : to wrap up the e2e conversation in the context of this PR, I suggest to file another PR (even a draft PR) kickstarting the changes to `sample-device-plugin` which you think you will need to enable e2e testing, and link back it here. This should be enough to demonstrate the work on e2e testing is ongoing and we will address in the coming weeks. I think we need (and want) to have e2e coverage as beta blocker anyway so we should be fine.

elezar commented:
@ffromani I have created a draft at https://github.com/kubernetes/kubernetes/pull/119280 to track the addition of the e2e tests. I will add more context to the PR in the coming days.

ffromani commented:
> @ffromani I have created a draft at #119280 to track the addition of the e2e tests. I will add more context to the PR in the coming days.

Perfect, thanks!

klueska reviewed (commented):
Thanks!

/lgtm
/approve

ffromani reviewed (commented):
/lgtm

to me as well

ffromani commented:
/cc @dims 

howdy! could you help with feature approval perhaps? This PR got LGTM and approval for all the other relevant areas

dims commented:
/approve

(pkg/features LGTM!)

ffromani commented:
@haircommander @elezar deferring the hold removal to you folks

elezar commented:
/unhold I believe there were still discussions around whether this needed a feature gate at the time. These have since been resolved.

ffromani commented:
> /unhold I believe there were still discussions around whether this needed a feature gate at the time. These have since been resolved.

note prow requires the commands to sit on their own line
```
/unhold

I believe there were...
```

ffromani commented:
/hold

restoring

ffromani commented:
mmmhh possibly the bot was faster than me. Should not be a big deal for reason stated in the comments, let's see.

ffromani commented:
interesting, the unit test failure looks legit, but it also look like just a ordering issue. Perhaps we use a data structure which does not guarantee the ordering? Just to not leave any stone unturned, is the order of devices ever relevant? (I expect no, but still)

elezar commented:
The ordering of devices should not matter. I will have a look at the
failures tomorrow first thing.

On Sun, Jul 16, 2023, 13:47 Francesco Romani ***@***.***>
wrote:

> interesting, the unit test failure looks legit, but it also look like just
> a ordering issue. Perhaps we use a data structure which does not guarantee
> the ordering? Just to not leave any stone unturned, is the order of devices
> ever relevant? (I expect no, but still)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/118254#issuecomment-1637062797>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHDMQLHXWY7DTKHFXOLWILXQPIF7ANCNFSM6AAAAAAYOSYAKE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

elezar commented:
> interesting, the unit test failure looks legit, but it also look like just a ordering issue. Perhaps we use a data structure which does not guarantee the ordering? Just to not leave any stone unturned, is the order of devices ever relevant? (I expect no, but still)

I had a look at the issue here, and is due to the non-deterministic iteration over [resources](https://github.com/kubernetes/kubernetes/blob/16534deedf1e3f7301b20041fafe15ff7f178904/pkg/kubelet/cm/devicemanager/pod_devices.go#L248) when constructing the options. I have updated the tests to check for matching elements of the individual slices instead.

ffromani commented:
/test pull-kubernetes-e2e-kind

klueska commented:
/lgtm
